src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.cs

```diff
@@ -0,0 +1,288 @@
+        "propfull", "propg", "sim", "struct", "svm", "switch", "try", "tryf", "unchecked", "unsafe", "using", "while");
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public InlineCompletionsHandler()
+    {
+    }
+
+    public override string Method => VSInternalMethods.TextDocumentInlineCompletionName;
+
+    public override bool MutatesSolutionState => false;
+
+    public override bool RequiresLSPSolution => true;
+
+    public override TextDocumentIdentifier? GetTextDocumentIdentifier(VSInternalInlineCompletionRequest request)
+    {
+        return request.TextDocument;
+    }
+
+    public override async Task<VSInternalInlineCompletionList> HandleRequestAsync(VSInternalInlineCompletionRequest request, RequestContext context, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(context.Document);
+
+        // First get available snippets if any.
+        var snippetInfoService = context.Document.Project.GetRequiredLanguageService<ISnippetInfoService>();
+        var snippetInfo = snippetInfoService.GetSnippetsIfAvailable();
+        if (!snippetInfo.Any())
+        {
+            return new VSInternalInlineCompletionList();
+        }
+
+        // Then attempt to get the word at the requested position.
+        var sourceText = await context.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+        var syntaxFactsService = context.Document.Project.GetRequiredLanguageService<ISyntaxFactsService>();
+        var linePosition = ProtocolConversions.PositionToLinePosition(request.Position);
+        var position = sourceText.Lines.GetPosition(linePosition);
+        if (!SnippetUtilities.TryGetWordOnLeft(position, sourceText, syntaxFactsService, out var wordOnLeft))
+        {
+            return new VSInternalInlineCompletionList();
+        }
+
+        // Find the snippet with shortcut text that matches the typed word.
+        var wordText = sourceText.GetSubText(wordOnLeft.Value).ToString();
+        if (!BuiltInSnippets.Contains(wordText, StringComparer.OrdinalIgnoreCase))
+        {
+            return new VSInternalInlineCompletionList();
+        }
+
+        var matchingSnippetInfo = snippetInfo.Single(s => wordText.Equals(s.Shortcut, StringComparison.OrdinalIgnoreCase));
+
+        var matchingSnippet = RetrieveSnippetFromXml(matchingSnippetInfo, context);
+        if (matchingSnippet == null)
+        {
+            return new VSInternalInlineCompletionList();
+        }
+
+        // We currently only support snippet expansions, the others require selection support which is not applicable here.
+        if (!matchingSnippet.SnippetTypes.Contains("Expansion", StringComparer.OrdinalIgnoreCase))
+        {
+            return new VSInternalInlineCompletionList();
+        }
+
+        var expansion = new ExpansionTemplate(matchingSnippet);
+
+        // Parse the snippet XML
+        var parsedSnippet = expansion.Parse();
```

Comment:
the parsed snippet needs to be cached.  Am working on that, but will do as a separate PR to reduce the size here

Note that we cannot cache the actual LSP snippet, as it is context dependent (class names / formatting), but we can cache the snippet parts we read from XML.

Reply:
We also may want to build specific error handling around this given it's reading user content. Also might make sense to do the same with `RetrieveSnippetFromXml` above given that reads content from disk

Reply:
I think I'm OK with the amount of error handling in there right now.  currently we do some basic checks ot make sure the file exists.  But otherwise if there are malformed snippets, imho it is ok to let those exceptions bubble up to the queue and let them be automatically reported and logged.

Happy to hear other thoughts though.

Reply:
I did find a few cases I should probably handle actually, so will update